### ISSUE TITLE
vmbackup flag -snapshot.deleteURL= not work 

### ISSUE THREAD

**Issue by ahfuzhang** [VictoriaMetrics/VictoriaMetrics#2055]
I use vm-backup like this:
```
./vmbackup-prod \
  -configFilePath="/etc/cos/config.ini" \
  -credsFilePath="/etc/cos/creds.ini" \
  -customS3Endpoint="https://cos.ap-guangzhou.myqcloud.com" \
  -dst="s3://bucket-125800000/tenant/Production/formal/vm-storage/0/2022-01-06" \
  -origin="s3://bucket-125800000/tenant/Production/formal/vm-storage/0/2022-01-06" \
  -storageDataPath="/data" \
  -httpListenAddr="0.0.0.0:8420" \
  -memory.allowedBytes=2GB \
  -concurrency=5 \
  -maxBytesPerSecond=20MB \
  -snapshot.createURL="http://127.0.0.1:8482/snapshot/create" \
  -snapshot.deleteURL="http://127.0.0.1:8482/snapshot/delete" &
```

after 6 days, there's no free disk space:
```
du $(pwd) -h -d 2
3.3G    /var/victoria-metrics/data/data/small/2022_01
285.5G  /var/victoria-metrics/data/data/small/snapshots   # snapshots still here
288.8G  /var/victoria-metrics/data/data/small
110.9G  /var/victoria-metrics/data/data/big/2022_01
323.9G  /var/victoria-metrics/data/data/big/snapshots   # snapshots still here
434.8G  /var/victoria-metrics/data/data/big
723.6G  /var/victoria-metrics/data/data
```

than I use this:
```
curl -G "http://127.0.0.1:8482/snapshot/delete_all"
```

and all is right.

so, I think flag -snapshot.deleteURL not works.

**Comment by denisgolius** [comment]
Hi @ahfuzhang !

1. Do you see any errors in vmstorage logs?
2. Please provide version of your vmstorage and vmbackup.

**Comment by ahfuzhang** [comment]
> Hi @ahfuzhang !
> 
> 1. Do you see any errors in vmstorage logs?
> 2. Please provide version of your vmstorage and vmbackup.

I don't see any error logs of vmstorage. I guess I use wrong config of s3, so vmbackup fails, and it doesn't remove snapshot when error happends.

bash-5.0# vmstorage-prod --version
vmstorage-20211202-134512-tags-**v1.70.0**-cluster-0-g0f9aaf9ce
bash-5.0# vmbackup-prod --version
vmbackup-20211202-133511-tags-**v1.70.0**-0-g7f2f26b25

**Comment by valyala** [comment]
>  I guess I use wrong config of s3, so vmbackup fails, and it doesn't remove snapshot when error happends.

Snapshots must be deleted when some errors happen with improperly configured `s3` according to the source code - https://github.com/VictoriaMetrics/VictoriaMetrics/blob/96dce63dbdd447fea085cad3b03d3d6ac9b675e3/app/vmbackup/main.go#L69 . This code is executed just before `vmbackup` finishes its work. It may fail executing only when `vmbackup` process was interrupted via `kill -9` or via similar unclean shutdown procedure.

**Comment by MorphBonehunter** [comment]
Today we first tried vmbackup and as we did something wrong with our s3 (minio) config the backup fails.
So as you said there should be an cleanup of the snapshot but i don't think this works.
I haven't coding skills, but in case the backup process fails in:https://github.com/VictoriaMetrics/VictoriaMetrics/blob/365d2ff0bfa2cffdfcf0be9d224715023f847d4e/app/vmbackup/main.go#L116
the `logger.Fatalf` is called...and if i understand it right Fatalf is an Print with an call to os.Exit.
This would make sense, as this is the last message we got on console bevor vmbackup ends.

**Comment by zekker6** [comment]
@MorphBonehunter This case is covered by the following: https://github.com/VictoriaMetrics/VictoriaMetrics/blob/365d2ff0bfa2cffdfcf0be9d224715023f847d4e/app/vmbackup/main.go#L83
`defer` keyword in go allows scheduling some function invocation when current function finishes, so this code block will be executed before `os.Exit` is called.

**Comment by MorphBonehunter** [comment]
@zekker6 are you sure:

> func Exit(code int): [Exit](https://pkg.go.dev/os#Exit) causes the current program to exit with the given status code. Conventionally, code zero indicates success, non-zero an error. The program terminates immediately; deferred functions are not run.

it's from [here](https://www.golinuxcloud.com/golang-handle-defer-calls-os-exit/#The_execution_of_the_defer_functions_is_not_always_guaranteed) but as i said, i have no idea about golang :)

As the defered func calls snapshot.Delete which in turn have an logger.Infof("Deleting snapshot %s", snapshotName), i would guess the snapshot delete should be visible in the output but that never happens as the last output on my case was "fatal   VictoriaMetrics/app/vmbackup/main.go:117        cannot create backup"

**Comment by zekker6** [comment]
@MorphBonehunter I think you are right here, good catch!
Thanks for pointing this out, I'm going to open a PR to fix this shortly.

**Comment by valyala** [comment]
The commit 9436ae3b07bfa53f6d8b687a1ea1f9a903c94252 should properly delete snapshot on backup error. This commit will be included in the next release. In the mean time it is possible to build `vmbackup` from this commit according to [this instruction](https://docs.victoriametrics.com/vmbackup.html#how-to-build-from-sources) and check whether it properly deleted snapshot on backup error.

**Comment by valyala** [comment]
The issue has been fixed in the [LTS release v1.87.4](https://github.com/VictoriaMetrics/VictoriaMetrics/releases/tag/v1.87.4). Closing it as fixed then.

**Comment by valyala** [comment]
FYI, the issue has been fixed also starting from [vmbackup v1.90.0](https://github.com/VictoriaMetrics/VictoriaMetrics/releases).
